prompting/rewards/web_retrieval.py

```diff
@@ -0,0 +1,112 @@
+        completion: str = "\n".join(response_event.completions)
+
+        if not completion:
+            return BatchRewardOutput(rewards=np.asarray([0]), timings=np.asarray([time.perf_counter() - timer_start]))
+
+        dataset_entry = DDGDatasetEntry.model_validate_json(reference)
+        search_term = dataset_entry.search_term
+        reference_content = dataset_entry.website_content
+
+        # URL and the content provided in the completion.
+        response_url, response_content = self._parse_response(completion)
+        # Content scraped from the URL provided in the completion.
+        response_url_scraped = self._extract_website_content(response_url)
+
+        # Similarity between search term and miner's scraped content.
+        search_response_sim = self._cosine_similarity(content1=search_term, content2=response_content)
+        # Similarity between reference conent and miner's scraped content.
+        # reference_response_sim = self._cosine_similarity(content1=reference_content, content2=response_content)
+
+        # If the URL provided in the completion is valid.
+        valid_url_score = 0
+        if response_url_scraped is not None:
+            valid_url_score = self._cosine_similarity(content1=response_content, content2=response_url_scraped)
+
+        # Similarity between search term and reference content.
+        search_reference_sim = self._cosine_similarity(content1=search_term, content2=reference_content)
+        response_reference_ratio = search_response_sim / search_reference_sim
+        score = (search_response_sim + valid_url_score) / 2
+        if abs(response_reference_ratio - 1) > _SEARCH_TERM_THRESH:
+            logger.info(
+                f"Reponse and reference scraped content relevance to the search term exceeds the threshold. "
+                f"Similarity: response = {search_response_sim:.2f}; reference = {search_reference_sim:.2f}"
+            )
+            score = 0
+        elif valid_url_score < _VALID_URL_SCORE:
+            logger.info(
+                f"Search term is not relevant to the scraped content, similarity {valid_url_score} < {_VALID_URL_SCORE}"
+            )
+            score = 0
+
+        return BatchRewardOutput(rewards=np.asarray([score]), timings=np.asarray([time.perf_counter() - timer_start]))
+
+    @staticmethod
+    def _extract_website_content(url) -> str:
+        website = trafilatura.fetch_url(url)
+        return trafilatura.extract(website)
+
+    @staticmethod
+    def _parse_response(completion: str) -> tuple[str, str]:
+        """Parse the completion text and extracts the URL and content.
+
+        Args:
+            completion: The text to parse.
+
+        Returns:
+            tuple: A tuple containing the URL (str or None) and the content (str).
+        """
+        url = None
+        lines = completion.strip().split("\n")
+
+        # First, try to find URL by parsing the last line that starts with "URL:".
```

Comment:
should we really try to parse them or should we just make them return a json like

```
[
{
   "website_content": ...,
   "website_url": ...,
   "relevant_section": ...,
},
]
```

for instance? That way they can give us url, the most relevant section + the whole website content without us having to do any awkward parsing

Reply:
There is no parsing as such, it just searches for URL and the end of response
However we can do JSON approach, which is less error prone